### PR TITLE
[RIP-43] Support Timing Messages with Arbitrary Time Delay

### PR DESCRIPTION
**Pull request by GenerousMan**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

Currently, RocketMQ's delay message feature only supports delayed delivery for specific time levels. Such delay message feature(only support specific levels of delay time) is no longer enough to support the flexible usage of rocketmq. Therefore, we need a delay message feature that supports arbitrary delay time.

I have written my proposal on the links below：
Google Doc: https://docs.google.com/document/d/1D6XWwY39p531c2aVi5HQll9iwzTUNT1haUFHqMoRkT0/edit?usp=sharing
Shimo:https://shimo.im/docs/gXqme9PKKpIeD7qo/

To optimize the commit history, this is a new PR. 
The old PR is :https://github.com/apache/rocketmq/pull/4558

## Brief changelog

1. Add classes to support the feature, such as  **TimerMessageStore**, **TimerWheel**, **TimerLog**, and **Slot** , etc.
2. Modify the logic of CommitLog.AsyncPutMessage(). Add two judgement(_isRolledTimerMessage(msg) / checkIfTimerMessage(msg)_) for timing messages. If the message is timing message, its topic and related properties will be converted. 
3. Modify the logic of BrokerController.start()/shutdown()/initialize(), etc. Add the **TimerMessageStore** into these processes.
4. Add some functions in related files(DefaultMessageStore, CommitLog, MappedFileQueue, MessageConst, etc).
5. Add unit tests for the feature.

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [✅] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [✅] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [✅] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [✅] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [✅] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Review by lizhanhui — APPROVED** [review]

**Review by RongtongJin — COMMENTED** [review]
1. There are benchmarks in the current example module, but I think it would be better to add the simplest example to tell users how to use timing messages with arbitrary time Delay. In addition, adding relevant documents is also a good choice
2. It would be better to add an IT test.
3. It is necessary to replace origin delay level message with new timer message in the POP consume. It can also be completed in the subsequent pull requests.
4. The current implementation needs to adapt to the slaveActingMaster mode. If no one is willing to do it, I am willing to do this in the subsequent pull requests later.

**Comment by fuyou001** [comment]
Conflicting files  @GenerousMan

File: broker/src/main/java/org/apache/rocketmq/broker/util/HookUtils.java [hunk 1]
```diff
@@ -122,10 +124,85 @@ public static PutMessageResult handleScheduleMessage(BrokerController brokerCont
...
+        }
+        return null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELIVER_MS) || null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_MS) || null != msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC);
+    }
+    private static PutMessageResult transformTimerMessage(BrokerController brokerController, MessageExtBrokerInner msg) {
+        //do transform
+        int delayLevel = msg.getDelayTimeLevel();
+        long deliverMs;
+        try {
+            if (msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC) != null) {
+                deliverMs = System.currentTimeMillis() + Integer.parseInt(msg.getProperty(MessageConst.PROPERTY_TIMER_DELAY_SEC)) * 1000;
```
Comment (rushsky518) [inline broker/src/main/java/org/apache/rocketmq/broker/util/HookUtils.java:174]: when set setDelayTimeSec(24 * 3600 * 3), integer overflows

Comment (GenerousMan) [reply broker/src/main/java/org/apache/rocketmq/broker/util/HookUtils.java]: thx, it has been resolved.

File: store/src/main/java/org/apache/rocketmq/store/MessageStore.java [hunk 2]
```diff
@@ -165,6 +166,10 @@ GetMessageResult getMessage(final String group, final String topic, final int qu
      */
     long getMinOffsetInQueue(final String topic, final int queueId);
 
+    TimerMessageStore getTimerMessageStore();
+
+    void setTimerMessageStore(TimerMessageStore timerMessageStore);
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/MessageStore.java:171]: It would be better. to add some comments for the interface.

**Review by RongtongJin — APPROVED** [review]

**Review by duhenglucky — APPROVED** [review]

File: store/src/main/java/org/apache/rocketmq/store/timer/TimerWheel.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                localBuffer.get().putLong(lastPos);
+            }
+        }
+    }
+
+    //check the timerwheel to see if its stored offset > maxOffset in timerlog
+    public long checkPhyPos(long timeStartMs, long maxOffset) {
+        long minFirst = Long.MAX_VALUE;
+        int firstSlotIndex = getSlotIndex(timeStartMs);
+        for (int i = 0; i < slotsTotal * 2; i++) {
```
Comment (AuroraTwinkle) [inline store/src/main/java/org/apache/rocketmq/store/timer/TimerWheel.java:174]: I noticed that `slotsTotal * 2` appears multiple times. Why do we need to multiply it by 2? I don't understand.

**Pull request merged**